File: src/main/java/net/flyingfishflash/loremlist/api/data/response/LrmListResponse.java
package net.flyingfishflash.loremlist.api.data.response;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.Set;
import java.util.UUID;
import kotlinx.datetime.Instant;
import net.flyingfishflash.loremlist.domain.lrmlist.LrmList;

public record LrmListResponse(
    UUID id,
    String name,
    String description,
    // "public" is a reserved word in Java and cannot be a record component identifier.
    @JsonProperty("public") boolean isPublic,
    String owner,
    Instant created,
    String creator,
    Instant updated,
    String updater,
    Set<LrmListItemResponse> items) {

  public static LrmListResponse fromLrmList(LrmList lrmList) {
    Set<LrmListItemResponse> items =
        lrmList.items().stream()
            .map(LrmListItemResponse::fromLrmListItem)
            .collect(java.util.stream.Collectors.toSet());
    return new LrmListResponse(
        lrmList.id(),
        lrmList.name(),
        lrmList.description(),
        lrmList.isPublic(),
        lrmList.owner(),
        lrmList.created(),
        lrmList.creator(),
        lrmList.updated(),
        lrmList.updater(),
        items);
  }
}

File: src/main/kotlin/net/flyingfishflash/loremlist/api/data/response/LrmListResponse.kt
package net.flyingfishflash.loremlist.api.data.response

import kotlinx.datetime.Instant
import kotlinx.serialization.EncodeDefault
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.Serializable
import net.flyingfishflash.loremlist.core.serialization.UUIDSerializer
import net.flyingfishflash.loremlist.domain.lrmlist.LrmList
import java.util.UUID

@Serializable
@OptIn(ExperimentalSerializationApi::class)
data class LrmListResponse(
  @Serializable(with = UUIDSerializer::class)
  val id: UUID,
  val name: String,
  val description: String? = null,
  @EncodeDefault val public: Boolean = false,
  val owner: String,
  val created: Instant? = null,
  val creator: String? = null,
  val updated: Instant? = null,
  val updater: String? = null,
  val items: Set<LrmItemResponse>,
) {
  companion object {
    fun fromLrmList(lrmList: LrmList): LrmListResponse {
      return LrmListResponse(
        id = lrmList.id,
        name = lrmList.name,
        description = lrmList.description,
        public = lrmList.public,
        owner = lrmList.owner,
        created = lrmList.created,
        creator = lrmList.creator,
        updated = lrmList.updated,
        updater = lrmList.updater,
        // TODO: address warning about useless null-safe operator
        items = lrmList.items?.map { LrmItemResponse.fromLrmItem(it) }?.toSet() ?: emptySet(),
      )
    }
  }
}
